## Q1/ Load the data, Women's Clothing E-Commerce Reviews.csv Download Women's Clothing E-Commerce Reviews.csv, into memory.

In [1]:
import pandas as pd
df = pd.read_csv("Womens Clothing E-Commerce Reviews.csv")
print(df.head())


   Unnamed: 0  Clothing ID  Age                    Title  \
0           0          767   33                      NaN   
1           1         1080   34                      NaN   
2           2         1077   60  Some major design flaws   
3           3         1049   50         My favorite buy!   
4           4          847   47         Flattering shirt   

                                         Review Text  Rating  Recommended IND  \
0  Absolutely wonderful - silky and sexy and comf...       4                1   
1  Love this dress!  it's sooo pretty.  i happene...       5                1   
2  I had such high hopes for this dress and reall...       3                0   
3  I love, love, love this jumpsuit. it's fun, fl...       5                1   
4  This shirt is very flattering to all due to th...       5                1   

   Positive Feedback Count   Division Name Department Name Class Name  
0                        0       Initmates        Intimate  Intimates  
1       

## Q2/ Please concatenate the Title, Review Text, Division Name, Department Name, and Class Name as a new feature of Reviews.

In [2]:

import pandas as pd

df = pd.read_csv("Womens Clothing E-Commerce Reviews.csv")
df['Reviews'] = df['Title'].fillna('') + ' ' + df['Review Text'].fillna('') + ' ' + df['Division Name'].fillna('') + ' ' + df['Department Name'].fillna('') + ' ' + df['Class Name'].fillna('')

print(df.head())

   Unnamed: 0  Clothing ID  Age                    Title  \
0           0          767   33                      NaN   
1           1         1080   34                      NaN   
2           2         1077   60  Some major design flaws   
3           3         1049   50         My favorite buy!   
4           4          847   47         Flattering shirt   

                                         Review Text  Rating  Recommended IND  \
0  Absolutely wonderful - silky and sexy and comf...       4                1   
1  Love this dress!  it's sooo pretty.  i happene...       5                1   
2  I had such high hopes for this dress and reall...       3                0   
3  I love, love, love this jumpsuit. it's fun, fl...       5                1   
4  This shirt is very flattering to all due to th...       5                1   

   Positive Feedback Count   Division Name Department Name Class Name  \
0                        0       Initmates        Intimate  Intimates   
1     

## Q3/Please clean the new feature of Reviews using regular expressions.

In [3]:
import pandas as pd
import re

df = pd.read_csv("Womens Clothing E-Commerce Reviews.csv")

df['Reviews'] = df['Title'].fillna('') + ' ' + df['Review Text'].fillna('') + ' ' + df['Division Name'].fillna('') + ' ' + df['Department Name'].fillna('') + ' ' + df['Class Name'].fillna('')

def clean_text(text):

    cleaned_text = re.sub(r'[^a-zA-Z0-9\s]', ' ', text)

    cleaned_text = cleaned_text.lower()

    cleaned_text = re.sub(r'\s+', ' ', cleaned_text).strip()
    return cleaned_text

df['Reviews'] = df['Reviews'].apply(clean_text)
print(df.head())


   Unnamed: 0  Clothing ID  Age                    Title  \
0           0          767   33                      NaN   
1           1         1080   34                      NaN   
2           2         1077   60  Some major design flaws   
3           3         1049   50         My favorite buy!   
4           4          847   47         Flattering shirt   

                                         Review Text  Rating  Recommended IND  \
0  Absolutely wonderful - silky and sexy and comf...       4                1   
1  Love this dress!  it's sooo pretty.  i happene...       5                1   
2  I had such high hopes for this dress and reall...       3                0   
3  I love, love, love this jumpsuit. it's fun, fl...       5                1   
4  This shirt is very flattering to all due to th...       5                1   

   Positive Feedback Count   Division Name Department Name Class Name  \
0                        0       Initmates        Intimate  Intimates   
1     

## Q4/Build an RNN model to forecast the Recommended IND based on Reviews using TensorFlow.

In [4]:
import pandas as pd
import numpy as np
import tensorflow as tf
from sklearn.model_selection import train_test_split
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences


df = pd.read_csv("Womens Clothing E-Commerce Reviews.csv")


df['Reviews'] = df[['Title', 'Review Text', 'Division Name', 'Department Name', 'Class Name']].apply(lambda x: ' '.join(x.dropna().astype(str)), axis=1)


X = df['Reviews'].values
y = df['Recommended IND'].values


tokenizer = Tokenizer()
tokenizer.fit_on_texts(X)
X_seq = tokenizer.texts_to_sequences(X)


max_sequence_length = 100
X_padded = pad_sequences(X_seq, maxlen=max_sequence_length, padding='post', truncating='post')


X_train, X_test, y_train, y_test = train_test_split(X_padded, y, test_size=0.2, random_state=42)


model = tf.keras.Sequential([
    tf.keras.layers.Embedding(input_dim=len(tokenizer.word_index) + 1, output_dim=64),
    tf.keras.layers.LSTM(units=64),
    tf.keras.layers.Dense(1, activation='sigmoid')
])


model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])


model.fit(X_train, y_train, epochs=5, batch_size=32, validation_data=(X_test, y_test))

loss, accuracy = model.evaluate(X_test, y_test)
print("Test Loss:", loss)
print("Test Accuracy:", accuracy)


Epoch 1/5
588/588 ━━━━━━━━━━━━━━━━━━━━ 31s 47ms/step - accuracy: 0.8258 - loss: 0.4798 - val_accuracy: 0.8189 - val_loss: 0.4711
Epoch 2/5
588/588 ━━━━━━━━━━━━━━━━━━━━ 27s 46ms/step - accuracy: 0.8211 - loss: 0.4491 - val_accuracy: 0.8189 - val_loss: 0.4389
Epoch 3/5
588/588 ━━━━━━━━━━━━━━━━━━━━ 27s 46ms/step - accuracy: 0.8198 - loss: 0.4190 - val_accuracy: 0.8246 - val_loss: 0.3595
Epoch 4/5
588/588 ━━━━━━━━━━━━━━━━━━━━ 28s 47ms/step - accuracy: 0.8943 - loss: 0.2612 - val_accuracy: 0.9019 - val_loss: 0.2384
Epoch 5/5
588/588 ━━━━━━━━━━━━━━━━━━━━ 28s 47ms/step - accuracy: 0.9321 - loss: 0.1768 - val_accuracy: 0.9072 - val_loss: 0.2255
147/147 ━━━━━━━━━━━━━━━━━━━━ 2s 14ms/step - accuracy: 0.9043 - loss: 0.2287
Test Loss: 0.22550635039806366
Test Accuracy: 0.9071945548057556


## Q5/Evaluate the model and check the model fit. Do you recommend this model? Please justify your answer.
 

In [5]:

loss, accuracy = model.evaluate(X_test, y_test)
print("Test Loss:", loss)
print("Test Accuracy:", accuracy)

147/147 ━━━━━━━━━━━━━━━━━━━━ 2s 14ms/step - accuracy: 0.9043 - loss: 0.2287
Test Loss: 0.22550635039806366
Test Accuracy: 0.9071945548057556


The model looks pretty good! With an accuracy of around 90.72%  and a loss of about 0.287, it seems like it's doing a solid job at figuring out whether people recommend products based on their reviews. So, yes, I'd recommend using this model. 